src/openzeppelin/governance/timelock/library.cairo

```diff
@@ -0,0 +1,505 @@
+
+        let (hash: felt) = hash_finalize{hash_ptr=pedersen_ptr}(state)
+
+        return (hash=hash)
+    end
+
+    func _get_calls_hash_array{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(
+        call_array_len: felt, 
+        call_array: TimelockCall*, 
+        calldata: felt*
+    ) -> (calls_hash_array: felt*):
+        alloc_locals
+
+        let (calls: Call*) = alloc()
+        _from_timelock_calls_to_calls(call_array_len, call_array, calldata, calls)
+        let calls_len = call_array_len
+
+        let (calls_hash_array: felt*) = alloc()
+        _hash_calls(calls_len, calls, calls_hash_array)
+
+        return (calls_hash_array=calls_hash_array)
+    end
+
+    func _from_timelock_calls_to_calls{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(
+        call_array_len: felt, 
+        call_array: TimelockCall*, 
+        calldata: felt*, 
+        calls: Call*
+    ):
+        if call_array_len == 0:
+            return ()
+        end
+
+        assert [calls] = Call(
+            to=[call_array].to,
+            selector=[call_array].selector,
+            calldata_len=[call_array].data_len,
+            calldata=calldata + [call_array].data_offset
+            )
+
+        _from_timelock_calls_to_calls(call_array_len - 1, call_array + TimelockCall.SIZE, calldata, calls + Call.SIZE)
+
+        return ()
+    end
+
+    func _hash_calls{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(
+        call_array_len: felt, 
+        call_array: Call*, 
+        hash_array: felt*
+    ):
+        if call_array_len == 0:
+            return ()
+        end
+
+        let (state: HashState*) = hash_init()
+        let (state: HashState*) = hash_update_single{hash_ptr=pedersen_ptr}(state, [call_array].to)
+
+        let (state: HashState*) = hash_update_single{hash_ptr=pedersen_ptr}(
+            state, [call_array].selector
+        )
+
+        let (state: HashState*) = hash_update{hash_ptr=pedersen_ptr}(
+            state, [call_array].calldata, [call_array].calldata_len
+        )
+
+        let (hash: felt) = hash_finalize{hash_ptr=pedersen_ptr}(state)
+        assert [hash_array] = hash
+
+        _hash_calls(call_array_len - 1, call_array + Call.SIZE, hash_array + 1)
+
+        return ()
+    end
+
+    func _execute_calls{
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(
+        id: felt, 
+        index: felt, 
+        calls_len: felt, 
+        calls: Call*
+    ):
+        alloc_locals
+
+        if calls_len == 0:
+            return ()
+        end
+
+        let this_call: Call = [calls]
+
+        call_contract(
+            contract_address=this_call.to,
+            function_selector=this_call.selector,
+            calldata_size=this_call.calldata_len,
+            calldata=this_call.calldata,
+        )
```

Comment:
I'm not sure about this. I think it makes sense to deconstruct the array to emit separate events, but not sure about making separate calls. Can you run some benchmarks to see if there's a big performance impact?